esp32/esp32_pcnt: Add PCNT class and Counter/Encoder shims in machine

jonathanhogg opened the pull request:
Provides pretty complete support for the ESP32 PCNT peripheral including control pin, second channel and events.

jonathanhogg commented:
Examples:

```python
from machine import Counter, Pin


# Simple counter:

def callback(counter):
    print("Counter value:", counter.value())


simple = Counter(0, pin=Pin(0, Pin.IN, pull=Pin.PULL_UP), rising=Counter.INCREMENT, threshold1=5, maximum=10, filter=1000)
simple.irq(handler=callback, trigger=Counter.IRQ_THRESHOLD1)
simple.start()


# Rotary encoder, 1X detection:

pin_a = Pin(12, Pin.IN, pull=Pin.PULL_UP)
pin_b = Pin(14, Pin.IN, pull=Pin.PULL_UP)

encoder = Counter(1)
encoder.init(pin=pin_a, falling=Counter.INCREMENT, rising=Counter.DECREMENT, mode_pin=pin_b, mode_low=Counter.HOLD)

# 2X detection:

encoder = Counter(1)
encoder.init(pin=pin_a, falling=Counter.INCREMENT, rising=Counter.DECREMENT, mode_pin=pin_b, mode_low=Counter.REVERSE)

# 4X detection:

encoder = Counter(1, filter=1000, minimum=-100, maximum=100)
encoder.init(0, pin=pin_a, falling=Counter.INCREMENT, rising=Counter.DECREMENT, mode_pin=pin_b, mode_low=Counter.REVERSE)
encoder.init(1, pin=pin_b, falling=Counter.DECREMENT, rising=Counter.INCREMENT, mode_pin=pin_a, mode_low=Counter.REVERSE)
```

IhorNehrutsa commented:
See https://github.com/micropython/micropython/pull/6639

jonathanhogg commented:
> See #6639

I did actually take a look at this implementation (I tried to hunt down all I could find), but I preferred to just expose the whole PCNT API through a single class rather than having a separate class for quadrature decoding – the latter being trivially implemented with the former.

That said – as per the comment in #5216 – I wrote this code for my own use case and opened this PR as an FYI for anyone interested in counters. It may make more sense for MicroPython in the general case to use a separate quadrature decoder class as this might map better to other ports (I see STM has its own way of doing quad decoding with counters). Rotary encoders are such a common use case that a specific API for them makes sense (kind of like the neopixel library).

dpgeorge commented:
Thanks, it looks interesting.

If this were to be accepted (as part of the machine module) it would need documentation and the possibility to eventually implement it for other ports, based on the docs.

jonathanhogg commented:
@dpgeorge: OK, you've sucked me in to writing some documentation for it – even though I said I wasn't going to try and see this through to mainline :wink:

### Notes:

- I wasn't clear on the form for documenting a class in the main library that is only (presently) supported on one port. It's hard to be certain what exactly the "common" API should be, so I've gone for a very limited API.
- I've documented the full ESP32 API in the ESP32 quick reference doc.
- There seems to be no consistency in the documents over English dialect. I have gone for the American spelling of "initialize" as – despite being a Brit – I've always found it safer to use American spelling for computing/electronics work and it's become a habit now.

I'm conscious of not wanting to step on the toes of the people who have tried to get a pulse counter PR into MicroPython before and I can see the value in @IhorNehrutsa's idea to have a separate class for quadrature decoding – although whether this ought to be in a pure-Python wrapper like the `neopixel` library is an open question.

That said, I'm committed to supporting this code as it's going into production hardware this year whether it goes mainline or not.

robert-hh commented on docs/library/machine.Counter.rst:
```diff
@@ -0,0 +1,74 @@
+    Stop the counter, i.e., finish monitoring the pin for transitions. The
+    current counter value will be retained while the counter is stopped.
+
+.. method::  I2S.value([value])
```
Most probably that should be: Counter.value([value])

jonathanhogg replied:
Oops! Cut-and-pasto…

jonathanhogg commented:
Leaving this PR around for reference and because I'm using this code in production devices. However, I'm not going to work on it any more.

I'm leaving @IhorNehrutsa and @robert-hh to come up with a common API for counters and encoders (#6639 and #7911 respectively) in the hope that one day we see it hitting mainline and I can throw this one in the bin.

robert-hh commented:
That's a little bit sad. A combination of your driver and the one of @IhorNehrutsa would be more complete. I just saw you made a great work on ADC, which improves it a lot. I hope it gets merged. It may take a while until the Encoder/Counter support will be in the main line, if ever. It does not seem to be a desperately sought feature.

jonathanhogg commented:
> That's a little bit sad. A combination of your driver and the one of @IhorNehrutsa would be more complete.

Yeah, I thought about this, but I think the main (only?) purpose of the modal functionality in the ESP32 PCNT API is to support encoders and this is going to be more understandable to a user wrapped in an Encoder class – and certainly the only way we'll see a cross-port API. The event callbacks are neat, but are going to be hard to merge in with the wide-counter support in @IhorNehrutsa's work and I feel like that's probably more useful to an end user.

> I just saw you made a great work on ADC, which improves it a lot. I hope it gets merged. 

Yes, I'm hoping so too. Support for ADC block 2 has been outstanding for ages and the ESP32 calibrated voltage API is really great.

> It may take a while until the Encoder/Counter support will be in the main line, if ever. It does not seem to be a desperately sought feature.

I actually had someone email me recently in desperate need of counter support on ESP32 and asking for me to cut them a custom build of my branch!

Having good PRs covering common API documentation, and MIMXRT and ESP32 implementations will be a great start. I feel like one more port would tip it over the line. STM32 seemed like an obvious choice, but hardware timer support  is bound up in the `pyb` module and that looks like a can of worms…

IhorNehrutsa commented:
@jonathanhogg
Would I add an IRQ support handler to #6639
for PCNT_EVT_THRES_0, PCNT_EVT_THRES_1 ?

jonathanhogg commented:
OK, so I didn't leave this alone after all…

I'm not happy continuing to use my API while the #8072 API exists, but I *am* comfortable continuing to use my implementation for the moment.

So I've moved my `Counter()` class into the `esp32` module as `PCNT()`. Then I've written two very slight shims around this class in the `machine` Python frozen module to adapt my old API to match the new API.

I've tweaked the IRQ handing of my `PCNT()` class to pass the triggered events instead of the object as the handler argument as this was needed to implement wide values.

jonathanhogg commented:
@harbaum, as per [my comment on the commit](https://github.com/jonathanhogg/micropython/commit/959eedd556789bd89e39fc13de596affd8639219#commitcomment-76790968):

It's always been my experience that callbacks with the mp scheduler execute on the next opcode boundary. So, if a hardware wrap occurs during the `self._pcnt.value()` call, I would expect the `_wrap()` method to have executed by the time that the follow-on `self._offset == offset` test executes.

That said, the pending flag is an interesting idea.

harbaum commented:
I'll give your version a try. I have seen this race condiction on my version and I should be able to reproduce it with yours.

jonathanhogg commented:
> I'll give your version a try. I have seen this race condiction on my version and I should be able to reproduce it with yours.

I'm not gonna lie, it's something of a hack, but I hadn't intended this code to go much further than my own use cases, which are pretty resilient to dodgy results from the counter.

I don't have any hardware with me today to test, but it ought to be easy enough to add some test code to reset the PCNT value inside the `.value()` method and then schedule a pretend call to the IRQ handler to say that it's wrapped.

harbaum commented:
This needs further debugging. I have set the minimum and maximum limits to +/-32 and the hw counter actually behaves accordingly. But _wrap is not called at all. Has this ever been tested?

jonathanhogg commented:
I tested it back when I first wrote the code, but haven't really been paying it much attention lately as other work has been sucking up my time.

harbaum commented:
I've spent an hour or two trying to figure out why the IRQ in your code doesn't seem to fire. You setup everything like I do ... but it doesn't work ... grumble ...

harbaum commented on docs/esp32/quickref.rst:
```diff
@@ -375,6 +375,25 @@ supported ADC resolutions:
+    from machine import Pin
+    from esp32 import PCNT
+
+    counter = PCNT(0, pin=Pin(2), rising=Counter.INCREMENT)     # create counter
```
AttributeError: type object 'Counter' has no attribute 'INCREMENT'

Shouldn't this be "Counter.UP"?

harbaum commented on ports/esp32/modules/machine.py:
```diff
@@ -0,0 +1,78 @@
+
+        def __init__(self, unit_id, *args, **kwargs):
+            self._pcnt = self._PCNT(unit_id, minimum=-32000, maximum=32000)
+            self._pcnt.irq(self._wrap, self._PCNT.IRQ_MINIMUM | self._PCNT.IRQ_MAXIMUM)
+            self.init(*args, **kwargs)
```
These two lines need to be swapped. Calling init re-initializes the counter and thus disables the IRQs again.

jonathanhogg replied:
That's what I get for making code changes when I should have just been rebasing…

harbaum commented:
And after fixing the IRQ order it becomes visible that your code is also suffering from the race condition. I am using this simple code with a rotary encoder on 34 and 35 and with the minimum/maximum in machine.py set to -32/32:

```
from machine import Pin, Encoder

enc = Encoder(0, Pin(34), Pin(35))

c = None
while True:
    v = enc.value()
    if c != v:
        print("Encoder value:", v)
        c = v
```

The resulting output is this:
```
...
Encoder value: -30
Encoder value: -31
Encoder value: 0
Encoder value: -32
Encoder value: -33
...
```

The hardware counter is reset to zero but the _wrap handler is called with enough delay to temporarily get a value of 0.

harbaum commented on docs/library/esp32.rst:
```diff
@@ -145,6 +145,98 @@ Constants
+    Reset and return is not atomic and so it is possible for a pulse to be
+    missed.
+
+.. method:: PCNT.irq(trigger=Counter.IRQ_ZERO, handler=None)
```
This should probably be PCNT.IRQ_ZERO

jonathanhogg replied:
Ta for catching these. I've found another and revised the `irq()` docs as well.

jonathanhogg commented:
> And after fixing the IRQ order it becomes visible that your code is also suffering from the race condition.

I'm gonna grab some hardware this morning and have a play later on to see if I can replicate and then come up with a neat solution.

harbaum commented:
> I'm gonna grab some hardware this morning and have a play later on to see if I can replicate and then come up with a neat solution.

Excellent. Looking forward to give it a test run.

jonathanhogg commented:
So I managed to replicate the race condition using a `Counter()` variant of your code, @harbaum (I have a button on my test board, but not an encoder).

I thought about this for a bit and then decided to make a more significant change to the `.irq()` handling. I like the pending events idea, but am not really happy with having `.value()` return a retry sentinel or raising a retry exception – it complicates all calling code for a rare instance. So I am now explicitly calling any `.irq()` handler in the `.value()` method if there are pending events.

To do this I've reworked the interrupt routine to set the object pending flag and then schedule a single static callback node. This callback function iterates over the PCNT objects calling handlers as necessary and clearing the pending events. I've had to enable the static callback scheduler for the **esp32** port to make this work.

This change guarantees that an event occurring before a call to `PCNT.value()` will be dealt with before that method returns and that the returned value will be as-of *after* that callback has executed. This means that I can be certain that `Counter._offset` has been updated before `PCNT.value()` returns. It is obviously still possible (and indeed slightly more likely) for `Counter._offset ` to have changed between retrieving it and `PCNT.value()` returning and so I still need the retry loop in the `Counter.value()` method. In my own testing this removes the race condition.

The downside is that this new code feels slightly more hairy to me. Indeed it's possible for `PCNT.value()` to be re-entered now if the `.irq()` handler calls it. I can't see any issues with this happening and have tried it in some test code without it causing problems, but it's an oddity to be aware of.

I have now put a great deal more though than I was expecting into this PR that I was planning to throw away… ;-)

harbaum commented:
Thanks a lot. Will test asap (probably on Monday).

harbaum commented:
My previous encoder test case now works perfectly. Thanks!

jonathanhogg commented:
Tidied up the documentation a little to explain the detail of `irq()`.

jonathanhogg commented:
Oops! Just had to fix a hilariously stupid bug in the logic for using `Counter.value()` to retrieve-and-replace the value.

jonathanhogg commented:
OK, the soft-reset crash was annoying me so I've finally fixed it: all PCNT objects are now de-initialised and freed on soft-reset.

harbaum commented:
Thanks a lot. The updated version still passes all my own test cases!

msamblanet commented:
I am testing the Encoder as implemented here in my project and the encoder is working fine.  However, when I use the encoder, I have an IRQ on another pin that I suddenly start receiving only once.  When I use the Encoder implementation at PR 7582, it works fine.

Details: ESP32S3 with encoder objects instantiated on 6/43, 38/39, and 2/4.  IRQ established on Pin 44 (the button of the 38/39 encoder).  When I click, I get the first IRQ but then never get any more when using this PR.  Replacing the firmware with a build of PR 7582 and it works as expected.

I also tried rebasing this PR to the MP master and no change in behavior.

I have limited time at the moment, so I have not yet dug in to see what the issue might be.

The encoders themselves are working nicely however...

jonathanhogg commented:
> I am testing the Encoder as implemented here in my project and the encoder is working fine. However, when I use the encoder, I have an IRQ on another pin that I suddenly start receiving only once. When I use the Encoder implementation at PR 7582, it works fine.
> 
> Details: ESP32S3 with encoder objects instantiated on 6/43, 38/39, and 2/4. IRQ established on Pin 44 (the button of the 38/39 encoder). When I click, I get the first IRQ but then never get any more when using this PR. Replacing the firmware with a build of PR 7582 and it works as expected.
> 
> I also tried rebasing this PR to the MP master and no change in behavior.
> 
> I have limited time at the moment, so I have not yet dug in to see what the issue might be.
> 
> The encoders themselves are working nicely however...

Hi @msamblanet 

Can you boil the problem down to the smallest example code you can for me? Perhaps just the setup of the button `Pin` and the `Encoder` would do. I'm away and not at any hardware for a few more days, but will test when I get back.

Also, what do you mean by "replacing the firmware with a build of PR 7582, it works fine"? This PR *is* #7582.

Cheers,

Jonathan

msamblanet commented:
@jonathanhogg 

RE: PR Number - sorry likely got PR numbers backwards.  That should have been PR #8766 - I am testing with both your implementation of Encoder and IhorNehrutsa's

I created a simplified program to demonstrate the issue but as I cut out stuff that was not needed, I found out the problem exists even without the encoder.

After testing a lot of Micropython downloads and local builds, I discovered that this works in 1.19.1 release but is broken in the 2022-08-29 nightly.  So it would seem this is a bug that was introduced in core Micropython between 1.19.1 release and whatever point in time your PR is based on.  Guessing an ESP32S3 specific bug since I don't think the ESP32 has a pin 44 (which is the one that is failing).

I'll go double check the issues there and report it if needed.

Thanks for your time to look at my issue...

- M

jonathanhogg commented:
> Thanks for your time to look at my issue...

No worries. I'm sorry that it's broken, but I guess happy that it wasn't me that broke it 😉

If you're bisecting, the last time I rebased this PR it was to 579f330508e4ba46f4a71193582c36bdf7aa56bd and v1.19.1 was 9b486340da22931cde82872f79e1c34db959548b – so I guess whatever has changed happened somewhere in there.

vahithosan commented:
Hi . Im testing this fork. Thank you. I have rotary encoder **(NOC-s500-2MWT)** and connected e**sp32-wroom-32d** with **pc817** optocoupler. I think I'm doing something wrong somewhere. Because the encoder misses in fast movements. 

And I have a question. I initially set the thresold1 and thresold2 values with rotary.init(). then what should i do to change it. If I use the init method again, the rotary.value() is reset.

I followed these steps to compile.

`export BUILD_DIR=$(pwd)`
```
git clone  https://github.com/jonathanhogg/micropython.git
cd $BUILD_DIR/micropython
git checkout 8cc6f0
```
```
cd $BUILD_DIR/micropython/
git clone --depth 1 -b v4.4 --recursive https://github.com/espressif/esp-idf.git
```
```
cd $BUILD_DIR/micropython/esp-idf
git checkout v4.4
git submodule update --init --recursive
./install.sh
. ./export.sh
```
```
cd $BUILD_DIR/micropython/mpy-cross
LC_ALL=en_US.UTF-8 make
```
```
cd $BUILD_DIR/micropython/ports/esp32
LC_ALL=en_US.UTF-8 make clean
LC_ALL=en_US.UTF-8 make submodules
LC_ALL=en_US.UTF-8 make
```

install firmware after boot:
```
MicroPython v1.12-2896-g8cc6f0c23-dirty on 2022-09-04; ESP32 module with ESP32
Type "help()" for more information.
```
My test file:

```
from machine import Pin
from esp32 import PCNT
import time


pin_a = Pin(16, Pin.IN)
pin_b = Pin(17, Pin.IN)
rotary = PCNT(0, pin=pin_a, falling=PCNT.INCREMENT, rising=PCNT.DECREMENT, mode_pin=pin_b, mode_low=PCNT.REVERSE)
rotary.init(channel=0, pin=pin_b, falling=PCNT.DECREMENT, rising=PCNT.INCREMENT, mode_pin=pin_a, mode_low=PCNT.REVERSE, threshold0=100, threshold1=300)
rotary.start()

def aa(p):
    print("100")
    rotary.irq(trigger=PCNT.IRQ_THRESHOLD1, handler=bb)
    
def bb(p):
    print("300")
    rotary.irq(trigger=PCNT.IRQ_THRESHOLD0, handler=aa)
    
rotary.irq(trigger=PCNT.IRQ_THRESHOLD0, handler=aa)
```

jonathanhogg commented:
> Hi . Im testing this fork. Thank you. I have rotary encoder **(NOC-s500-2MWT)** and connected e**sp32-wroom-32d** with **pc817** optocoupler. I think I'm doing something wrong somewhere. Because the encoder misses in fast movements.

Hi @vahithosan,

From what I can see, you're setting up the `PCNT` object as I would expect. However, the `.init()` call you are making after creating the object reconfigures the counter the other way around from the way that you've configured it in the constructor:

> ```python
> rotary = PCNT(0, pin=pin_a, falling=PCNT.INCREMENT, rising=PCNT.DECREMENT, mode_pin=pin_b, mode_low=PCNT.REVERSE)
> rotary.init(channel=0, pin=pin_b, falling=PCNT.DECREMENT, rising=PCNT.INCREMENT, mode_pin=pin_a, mode_low=PCNT.REVERSE, threshold0=100, threshold1=300)
> ```

You can just do this as:

```python
rotary = PCNT(0, pin=pin_a, falling=PCNT.INCREMENT, rising=PCNT.DECREMENT, mode_pin=pin_b, mode_low=PCNT.REVERSE, threshold0=100, threshold1=300)
```

> And I have a question. I initially set the thresold1 and thresold2 values with rotary.init(). then what should i do to change it. If I use the init method again, the rotary.value() is reset.

So this is a misfeature of my implementation at the moment: you can only reconfigure a `PCNT` channel all at once and that also resets the value. Ideally, you would call the `.init()` method with just the threshold keyword arguments and these would be applied in isolation. I should make this change, but haven't had the time (or need to be honest) to work on this PR for a little while.

Also, note that you can configure both interrupt events at the same time and differentiate between them from the flags argument that the handler is passed:

```python
def threshold_handler(trigger):
    if trigger & PCNT.IRQ_THRESHOLD0:
        print("100")
    elif trigger & PCNT.IRQ_THRESHOLD1:
        print("300")

rotary.irq(trigger=PCNT.IRQ_THRESHOLD0 | PCNT.IRQ_THRESHOLD1, handler=threshold_handler)
```

To return to the original question, though, I don't know what might be causing the counter to miss pulses. It might be worth adding the `filter` argument in case bounce on the A and B lines is being interpreted as backward steps of the encoder.

msamblanet commented:
@vahithosan based on my experiences, these encoders can be quite bouncy and this can cause strange behavior.  In my testing, I’ve been using maximum filter time on the ESP and once set, been seeing great results.  I have a 100 step per rotation encoder and when I turn by hand at about 6 rotations a second (so 2400 pulses per second), I have nothing drop.

jonathanhogg commented:
Yeah, I can confirm from my own applications that the PCNT hardware is solid at a million pulses per second – although I'm using a simple pulse counter not quadrature encoded pulses.

jonathanhogg commented:
So I felt inspired by the renewed interest in this PR to finally rewrite the initialisation so that it supports partial reconfiguration of a PCNT unit. I've updated the docs, but the key notes are:

- the `PCNT()` constructor now only fully reinitialises a PCNT unit the first time you call it with a particular unit `id`
- all keyword arguments are passed straight to `.init()`
- `.init()` (re)configures the unit in the most granular way that the IDF supports, leaving anything unspecified alone if possible; it also no longer resets the counter value (if it can) and no longer disables any `.irq()` handler
- call `.deinit()` if you want to return a PCNT unit to its initial state

You can partially reconfigure the PCNT unit with the following individual/groups of keyword arguments:

- `pin` (and `mode_pin`) to change just the bound pin(s)
- `rising`, `falling`, `mode_low` and `mode_high` to change the counting logic (specifying any one of these will reconfigure the logic with any missing keywords taking their default values)
- `filter` (with 0 meaning no filter)
- `value` (set to 0 to force a counter reset)
- `minimum`
- `maximum`
- `threshold0`
- `threshold1`

Unfortunately, you can't set any of the limit/event values (`minimum`, `maximum`, `threshold0` and `threshold1`) without resetting the counter value, so adding any of these arguments implies `value=0`.

I've rebased to the current micropython tip. I see that I'm failing a code formatting check, but I'll leave that be until you guys have been able to retest with this and confirmed that I've not broken your applications.

msamblanet commented:
@jonathanhogg thank you very much for this.  I will test as soon as I can but I am out of the country till the end of September but will test it and report back then.

jonathanhogg committed:
esp32/modules/machine.py: Add Counter and Encoder shims.

Add thin Python shims around `esp32.PCNT` to implement the
`machine.Counter` and `machine.Encoder` classes (as per #8072 API).

jonathanhogg commented:
Right. More changes to tick off the last item on my to-do list: the `machine.Counter` and `machine.Encoder` APIs do a better job of hiding the underlying PCNT hardware.

You can now create `Counter(0, ...)` and `Encoder(0, ...)` and these will use *different* PCNT units. The id passed to the constructors is completely arbitrary, so `Counter(1337, ...)` is just as valid.

To support this, I've tweaked the `PCNT()` constructor so that it can now be called without a unit `id` positional argument to get the next free PCNT unit – and by "free" I mean one that hasn't already been allocated. Note that there is no way to deallocate a PCNT object.

`Counter(id)` and `Encoder(id)` will return the *existing* object if one with that id has already been created. So `Counter(1) is Counter(1)`. The respective `.init()` methods will only be called if additional arguments are supplied to the constructor.

jonathanhogg committed:
docs/esp32: Add PCNT documentation.

Document the new `esp32.PCNT` class for hardware pulse counting.

jonathanhogg commented:
(Oops! Forgot to update the `esp32` library docs.)

harbaum commented:
Excellent. The id handling between Counter and Encoder was one of the issues I had with the other PR. A beginner would never understand why two seemingly unrelated things would affect each other. Now this was less a problem with your code as here it's visible that Counter and Encoder are both visibly derived from PCNT so some relation my be obvious.

Anyway, I like the fact that now they aren't affecting each other.

The downside now is: This now behaves differently than all other hardware classes which do honor the id. 

What about allowing non negative IDs and honoring them and then also allow negative IDs or 'None' as ID to trigger the new automatic assignment?

jonathanhogg commented:
> Anyway, I like the fact that now they aren't affecting each other.
> 
> The downside now is: This now behaves differently than all other hardware classes which do honor the id.

I suppose my goal is to make code as portable as possible. While there are few implementations of the `machine.Counter/Encoder` API at the moment, doing this allows code like:

```python
counter = Counter(0, Pin(2), filter_ns=10000)
encoder = Encoder(0, Pin(12), Pin(14), filter_ns=10000)
```

to just work without worrying about the underlying hardware and without placing any constraints on other ports. The PCNT units are completely interchangeable in ESP32, so which id is which really doesn't matter. However, if someone is really concerned about the detailed hardware operation then they can just use the `esp32.PCNT` API instead.

> What about allowing non negative IDs and honoring them and then also allow negative IDs or 'None' as ID to trigger the new automatic assignment?

I'm not a fan of negative ids as they aren't common elsewhere in MicroPython, already have a connotation of meaning a software implementation rather than hardware, and make the code less portable (since other ports are unlikely to follow the same convention).

`None` doesn't work as the important thing with the ids is that they are stable – once you've created `Counter(0)`, `Counter(0)` will always return the same object using the same underlying PCNT unit and `Counter(1)` will be a different stable unit.

All that said, I'm always open to being argued out of a design decision 😉

IhorNehrutsa commented:
Everyone knows the common expression: "truth is born in a dispute", but few people remember that this statement has a continuation - “it dies in that dispute”.

I am deeply disappointed with the attitude of MicroPython holders towards Encoder/Counter

jonathanhogg committed:
docs/library/machine: Add documentation of Counter and Encoder classes.

Add documentation for the core Counter/Encoder API as currently
implemented by the esp32 port. This documentation is a simplification of
that in PR #8072.

vahithosan commented:
Hi. I cant build for esp32-wroom-32d Generic.
```
>>> from esp32 import PCNT
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: can't import name PCNT
>>> 
```

I edited  mpconfigboard.h in GENERIC 
```
#define MICROPY_HW_BOARD_NAME               "ESP32 module"
#define MICROPY_HW_MCU_NAME                 "ESP32"

#define MICROPY_PY_ESP32_PCNT               (0)
```


I tried several combinations
7582 pr - esp-idf 4.2.2
7582 pr - esp-idf 4.4
(by manually editing files)
micropython version 1.18 -  esp-idf 4.2.2
micropython version 1.18 -  esp-idf 4.4
micropython version 1.19 -  esp-idf 4.2.2
micropython version 1.19 -  esp-idf 4.4

```

MicroPython 6f429bd-dirty on 2022-09-12; ESP32 module with ESP32
Type "help()" for more information.
>>> 
>>> from esp32 import PCNT
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ImportError: can't import name PCNT
>>> 
>>> import esp32
>>> help(esp32)
object <module 'esp32'> is of type module
  __name__ -- esp32
  wake_on_touch -- <function>
  wake_on_ext0 -- <function>
  wake_on_ext1 -- <function>
  wake_on_ulp -- <function>
  gpio_deep_sleep_hold -- <function>
  raw_temperature -- <function>
  hall_sensor -- <function>
  idf_heap_info -- <function>
  NVS -- <class 'NVS'>
  Partition -- <class 'Partition'>
  RMT -- <class 'RMT'>
  ULP -- <class 'ULP'>
  WAKEUP_ALL_LOW -- False
  WAKEUP_ANY_HIGH -- True
  HEAP_DATA -- 4
  HEAP_EXEC -- 1
>>> 
```

jonathanhogg commented:
> I edited mpconfigboard.h in GENERIC
> 
> ```
> #define MICROPY_HW_BOARD_NAME               "ESP32 module"
> #define MICROPY_HW_MCU_NAME                 "ESP32"
> 
> #define MICROPY_PY_ESP32_PCNT               (0)
> ```

The PCNT class is built by default for ESP32, setting `MICROPY_PY_ESP32_PCNT` to 0 in your board config turns it off. I only do that for the C3 modules because they don't have PCNT hardware.

vahithosan commented:
> > I edited mpconfigboard.h in GENERIC
> > ```
> > #define MICROPY_HW_BOARD_NAME               "ESP32 module"
> > #define MICROPY_HW_MCU_NAME                 "ESP32"
> > 
> > #define MICROPY_PY_ESP32_PCNT               (0)
> > ```
> 
> The PCNT class is built by default for ESP32, setting `MICROPY_PY_ESP32_PCNT` to 0 in your board config turns it off. I only do that for the C3 modules because they don't have PCNT hardware.

Thanks. I will try

vahithosan commented:
> The PCNT class is built by default for ESP32, setting MICROPY_PY_ESP32_PCNT to 0 in your board config turns it off. I only do that for the C3 modules because they don't have PCNT hardware.


Thanks for helping. (My English is not good. Sorry)
I did some work on PCNT.
I have an optical Rotary (NOC-s500-2MWT). Connected with ESP32 via PC817 optocoupler
Micropython version:
```
>>> import platform
>>> platform.platform()
'MicroPython-1.19.1-xtensa-IDFv4.4-with-newlib3.3.0'
```
But first I measured the operating speed of the using the other ESP32 RMT.
I'm sending 2 pulses. I expect the value to increase by 2

```
r = esp32.RMT(0, pin=Pin(2), clock_div=80)
r.write_pulses((1, 1, 1, 1), 0)
```
1 alternance 1 microsecond. 500khz
```
>>> counter.value()
0
>>> counter.value()
2
```
500 Khz worked very well.

then I measured again using PC817.
```
>>> r = esp32.RMT(0, pin=Pin(18), clock_div=80)
>>> counter.value()
0
```
1 alternance 500us
1kHz
```
>>> r.write_pulses((500,500,500,500),0)
>>> counter.value()
2
```
1 alternance 50us
10kHz
```
r.write_pulses((50,50,50,50),0)
>>> counter.value()
4
```
1 alternance 5us
100kHz missed 
```
r.write_pulses((5,5,5,5),0)
>>> counter.value()
5
```
Over 10kHz missed begin (25khz for example)

I wrote earlier that the encoder misses fast movements. The reason is the optocoupler.

I also tested it as an encoder

```
from machine import Pin
from esp32 import PCNT

pin_a = Pin(16, Pin.IN)
pin_b = Pin(17, Pin.IN)

rotary = PCNT(0, pin=pin_a, falling=PCNT.INCREMENT, rising=PCNT.DECREMENT, mode_pin=pin_b, mode_low=PCNT.REVERSE, threshold0=100, threshold1=300)
rotary.init(channel=1, pin=pin_b, falling=PCNT.DECREMENT, rising=PCNT.INCREMENT, mode_pin=pin_a, mode_low=PCNT.REVERSE)
rotary.start()
```
There was no missed pulses

irq() was not disable  when I updated with init().
```
>>> rotary.irq()
>>> rotary.value()
1399
```
The counter was not reset when I changed the filter value.
```
>>> rotary.init(filter=500)
>>> rotary.value()
1399
```
When I re-entered the threshold values with init(), the counter reset. It would be nice if the counter does not reset when the threshold values change.
```
>>> rotary.value()
682
>>> rotary.init(threshold0=200,threshold1=500)
>>> rotary.value()
0

```
I think it works well. Thanks

jonathanhogg commented:
> I wrote earlier that the encoder misses fast movements. The reason is the optocoupler.

OK, that makes sense – the PC817 datasheet suggests that the rise/fall time is around 4µs.

> When I re-entered the threshold values with init(), the counter reset. It would be nice if the counter does not reset when the threshold values change.

Yes, it would be nice. I spent quite a bit of time debugging this and unfortunately it seems to be a limitation of either the hardware or the IDF. Changes to any of the limits or threshold values just don't apply until the counter is cleared, so I decided that it made sense to force a reset whenever you set one of these to avoid user confusion.

Generally, for quadrature decoding the `machine.Encoder()` class is much more straightforward to use than the `PCNT()` class, but it doesn't support IRQs at the moment.

I'm curious as to what you plan to use the thresholds for in your application and why you would want to change them – perhaps there is another solution I could work on…

vahithosan commented:
> I'm curious as to what you plan to use the thresholds for in your application and why you would want to change them – perhaps there is another solution I could work on…

Thank you for your interest
There is a plate that goes up and down with the Electric Motor like an elevator. There are up and down limit switches that limit the movement of this plate. The encoder reads the movement of the plate. The system takes the minimum and maximum values with up and down limit switches in its first operation. Then the user enters the lower and upper values into the system. The system goes up and down in this range.  Like this

https://www.youtube.com/shorts/EvN41q2TsDg

If I change the threshold values while the system is running, the encoder value will reset  and the system will work incorrectly.
When the threshold value changes, it is necessary to restart the system again.

Maybe it can be solved by inserting value with rotaty.value() ?

I have applied using pin.irq, but I cannot catch the Encoder pulses when the motor is running fast.  I followed your work and IhorNehrutsa's work.

jonathanhogg commented:
> Maybe it can be solved by inserting value with rotaty.value() ?

Very interesting. Yes, I suspect that the solution is to read the current value and then offset the thresholds by this.

Something like:

```python
rotary = PCNT(0)
rotary.init(channel=0, ...)
rotary.init(channel=1, ...)
rotary.irq(switch_direction, PCNT.IRQ_THRESHOLD0 | PCNT.IRQ_THRESHOLD1)
base = 0

# always add base when reading current value:
position = rotary.value() + base

# figure out lower and upper values using position...

# set new thresholds, moving base to account for value reset:
base -= rotary.value()
rotary.init(threshold0=lower+base, threshold1=upper+base)
```

Doing this in my head without working out the maths, so I may have a sign wrong somewhere…

You'd be OK as long as `base` didn't drift far enough to push the thresholds out of the -32768...32767 PCNT value range as that would cause an exception and you'd have to do a system reset.

You've got the possibility of losing pulses between the `base` update and the `.init()` threshold reset, but this is likely not going to be severe unless the system is moving very quickly.

jonathanhogg commented:
Rebased to 13c4470fd04981fce3fd29771e1d973febbe7d60 and updated for new type declaration method.

jonathanhogg commented:
> Doing this in my head without working out the maths, so I may have a sign wrong somewhere…

So, looking at this again @vahithosan, I did get the signs wrong. It should have been:

```python
rotary = PCNT(0)
rotary.init(channel=0, ...)
rotary.init(channel=1, ...)
rotary.irq(switch_direction, PCNT.IRQ_THRESHOLD0 | PCNT.IRQ_THRESHOLD1)
base = 0

# always add base when reading current value:
position = rotary.value() + base

# figure out lower and upper values using position...

# set new thresholds, moving base to account for value reset:
base += rotary.value()
rotary.init(threshold0=lower-base, threshold1=upper-base)
```

vahithosan commented:
> > Doing this in my head without working out the maths, so I may have a sign wrong somewhere…
> 
> So, looking at this again @vahithosan, I did get the signs wrong. It should have been:
> 
> ```python
> rotary = PCNT(0)
> rotary.init(channel=0, ...)
> rotary.init(channel=1, ...)
> rotary.irq(switch_direction, PCNT.IRQ_THRESHOLD0 | PCNT.IRQ_THRESHOLD1)
> base = 0
> 
> # always add base when reading current value:
> position = rotary.value() + base
> 
> # figure out lower and upper values using position...
> 
> # set new thresholds, moving base to account for value reset:
> base += rotary.value()
> rotary.init(threshold0=lower-base, threshold1=upper-base)
> ```

Thank you. I haven't tried it this way. I'll take the plate to the bottom switch every time Thresold changes. The moment the down switch sees it, it will get the new threshold values with init(). also the rotary.value() will be reset.

But I will try your method as well.

Should thresold0 and threshold1 be between -32000 +32000?

jonathanhogg commented:
> Should thresold0 and threshold1 be between -32000 +32000?

Yes, they need to be within the 16-bit counter value range.

Resetting at a known switch point is always going to be the most reliable method.

jonathanhogg commented:
I'm marking this PR as "ready for review" as I'm using it as-is in production hardware and I'm happy for other people to use it if they want to. I don't see myself making any API or functionality changes in the near term, but will continue to rebase it every now and then.

IhorNehrutsa commented on ports/esp32/esp32_pcnt.c:
```diff
@@ -0,0 +1,450 @@
+            mp_raise_ValueError(MP_ERROR_TEXT("can only set value to 0"));
+        }
+    }
+    while (true) {
+        check_esp_err(pcnt_get_counter_value(self->unit, &value));
+        if (self->pending && self->handler != MP_OBJ_NULL) {
+            esp32_pcnt_run_irq_handlers(NULL);
+        } else {
+            break;
+        }
+    }
```
I'm not sure, looks like a deadlock if value() is called from an isr.
I'm not sure, looks like a deadlock if value() is called from an isr.
I'm not sure, looks like a deadlock if value() is called from an isr.
I'm not sure, looks like a deadlock if value() is called from an isr.
I'm not sure, looks like a deadlock if value() is called from an isr.
```
def cb():
     print(encoder.value())
```

jonathanhogg replied:
The pending flag is cleared before entering the callback, so the callback will only re-enter itself via a call to `.value()` if another interrupt occurs before the call to `.value()`. Even in that case, it should only deadlock if the callback can't be re-entered for some reason (i.e., it acquires some non-reentrant lock).

`.value()` itself is safe to re-enter and I've tested this exact code path. If you suffered some kind of interrupt storm (perhaps by misconfiguring the limits on very fast moving counter) then you might run out of stack, but it won't deadlock.

There's an argument to not allowing `.value()` to trigger callbacks if it is called from within a callback, but detecting that would require additional careful use of state. At the moment, I'm more inclined to re-write the interrupt stuff completely to support both soft and hard callbacks, but that will have to wait until I have more time…

cederom commented:
Hello world :-) Any chances that this code will be merged into a release? :-)

jonathanhogg commented:
> Hello world :-) Any chances that this code will be merged into a release? :-)

Given I've been working on it – in one form or another – for 18 months and mine is neither the first nor the only implementation, I'd not hold my breath if I was you…

I am committed to continuing to maintain this patch as I use it in a lot of my hardware projects, so feel free to keep using it. I know having to build your own firmware is a bore. I ought to look into automating that.

harbaum commented:
For the records: My custom hardware also exclusively uses images including this patch. It's really useful.

cederom commented:
Thanks @jonathanhogg for sharing your time and code :-)

I need to read rotary encoder and I like to rapid prototype with MicroPython so I am looking around if its gonna be MP or RTOS :-)

Firmware compilation is not a problem (I work on FreeBSD where MP does not build out of the box yet but I can always run a linux vm).

Having that feature within a release would be cool. It would enable using industrial rotary control and the small ones for use with UI.

Maybe different implementations could co-exist next to each other for users to select which one to use :-)

Do you know is it possible to use raw `PCNT` with `v1.19.1` or a custom firmware build is also required?

I would prefer to use `PCNT` not the `IRQ` approach.. but I also wanted to use ESP32-C3 and it turns out it has no `PCNT` :-( The impulse speed would go around 1MHz maybe more so a peripheral seems a must..  will have to use ESP32 then well :-)

Project owners please add this code to the nearest release :-)

jonathanhogg commented:
> Do you know is it possible to use raw `PCNT` with `v1.19.1` or a custom firmware build is also required?
> 
> I would prefer to use `PCNT` not the `IRQ` approach.. but I also wanted to use ESP32-C3 and it turns out it has no `PCNT` :-( The impulse speed would go around 1MHz maybe more so a peripheral seems a must.. will have to use ESP32 then well :-)

There is no ESP32 PCNT support in the mainline releases. My patch here adds a `PCNT` class into the `esp32` module and then additionally adds simpler `Counter` and `Encoder` classes in the `machine` module (implemented as pure Python wrappers around `PCNT`).

The PCNT peripheral is really very useful: I routinely count cycle-accurate pulse streams at 1Mhz. Doing that with IRQs is much harder!

cederom commented:
Why is the code not yet in the mainline? o_O

Would that be possible to create python modules for `PCNT`, `Counter` and `Encoder` so they run on current `v1.19.1` release or more internals are required than that? :-P

robert-hh commented:
> Why is the code not yet in the mainline? o_O

Mostly because a common interface for all ports does not have been defined. And for ESP32 there are three PR providing support for Encoder & Counter. But there may be hope. There seems to be the intentions that until mid 2023 all open PRs will be considered, which may include closing them.

cederom commented:
> > Why is the code not yet in the mainline? o_O
> 
> Mostly because a common interface for all ports does not have been defined. And for ESP32 there are three PR providing support for Encoder & Counter. But there may be hope. 

Cool, sounds like the code is ready for some time, API needs to be set, code update follows, and voila, I keep my fingers crossed.. but also sounds like no MP use for my testing at the moment.. thanks @robert-hh for clarification :-)

jonathanhogg commented:
> Why is the code not yet in the mainline? o_O

It doesn't seem to be a particular priority for the maintainers.

> Would that be possible to create python modules for `PCNT`, `Counter` and `Encoder` so they run on current `v1.19.1` release or more internals modifications are required than that? :-P

The `PCNT` code uses the Espressif IDF APIs and so is implemented in C and needs to be compiled into the firmware.

cederom commented:
> > Why is the code not yet in the mainline? o_O
> 
> It doesn't seem to be a particular priority for the maintainers.

Sad. All code seems to be here. Only API needs to be clarified, then probably minor code adjustments. This rules out MicroPython from many current use cases.. once started the RTOS project there is no need to come back to MicroPython in that case :-(

harbaum commented:
PCNT is very ESP specific.

Encoder and Counter on the other hand are independent from that and don't even have to be frozen into MP as they can be added in pure python on top of PCNT later.

I don't understand why the integration of PCNT needs to be delayed until Encoder and Counter are stable.

cederom commented:
From user perspective, I found working Encoder and CAN code with examples running on NuttX RTOS that builds and flashes under a minute so I will switch to that platform, big delays of useful verified stuff into a release is a big disadvantage of MicroPython project, please reconsider..

IhorNehrutsa commented:
Unfortunately, this is not the first and not the only case of delays in solving problems in Micropython. Unfortunately for all... :(

X-Ryl669 commented:
Just a remark here, PCNT is also useful for Flow meters, where neither Counter or Encoder are required. So the argument about waiting for Encoder or Counter to be merged first seems like it's counter intuitive.

X-Ryl669 commented:
Would it be possible to implement this as a [native module](https://docs.micropython.org/en/latest/develop/natmod.html) or is it relying on some symbols that's not in default micropython firmware on ESP32 ?

IhorNehrutsa commented:
@X-Ryl669 
> Just a remark here, PCNT is also useful for Flow meters....

May you post some link of Flow meters timing chart?
Or Flow meters models.
Thanks.

X-Ryl669 commented:
flow meters are very slow, compared to an rotary/axial encoder. The typical frequency will be in the hundred of Hz, not kHz. Their duty cycle is completely random either, since they are based on a magnet + hall effect sensor + Schmidt Trigger.  
You can have very short pulse, but with a large (random) period between them, or the opposite.

However, they are observing a physical system with friction and resistance so their answer is very non linear:
- When the flow is very low, the turbine or blade will not move due to the fluid going around the obstacle instead of pushing it, 
- When the flow is very high, the turbine will dampen the flow itself (the back of the blade will hit the flow and give a resistance). 
 
So one must apply different calibration table based on what regime they are in. This means that the number of pulse isn't very important, but the number of pulse per second is, since it allows to find out what regime they are in, and then you can interpret the pulse count. So basically, you'll have to sort the pulses in different buckets, one bucket for each (calibrated) regime.

So an implementation that's based on a Pin input IRQ and a getTickUs() will not work well, since you always have the issue of not knowing when the IRQ triggered (and the time between the handler is executing the get time function). Worse: you'll usually zero your counter in the handler and store the "last" time in a variable too only if some period elapsed. 
This delay is the issue since any pulse getting in while the IRQ is processing will be lost or its timing unsure.
You can't sleep while doing this either, so this is also quite power inefficient.

With a PCNT, and a Timer, you'll have an interrupt for the timer at a very precise interval, where you can read the pulse counter and deduce the regime you're in and the accumulated flow/pulses, while still having the PCNT running and counting new pulses. Also, you can sleep while the HW is counting the pulses.

jonathanhogg commented:
@X-Ryl669:

> Would it be possible to implement this as a [native module](https://docs.micropython.org/en/latest/develop/natmod.html) or is it relying on some symbols that's not in default micropython firmware on ESP32 ?

It relies on symbols that are not available from a native module. The IDF PCNT driver isn't even compiled in without being explicitly linked to.

There is actually surprisingly little that you can do from a native module as they require an explicit MicroPython lookup table for linking to anything at all. On the ESP32 you can't even do floating point division in a native module. Believe me, I've bounced up against these limitations often.

> Just a remark here, PCNT is also useful for Flow meters, where neither Counter or Encoder are required (we need to assert the pulse per time period, not the count of pulses, and ideally, we don't want to run an IRQ per pulse). So the argument about waiting for Encoder or Counter to be merged first seems like it's counter intuitive to me.

Honestly, I don't think that anything other that a lack of interest / core-developer time is holding up mainlining a PCNT implementation. There have been numerous runs at it over the last couple of years and if a particular API was or wasn't the problem that that would have been indicated and resolved by now.

As noted before, I'm still committed to maintaining this PR. I've been working on higher-level stuff for the last few months, but I'll rebase this to 1.20 in a moment.

IhorNehrutsa commented:
@jonathanhogg @harbaum 
Have you tried changing threshold1 like here?
```
from machine import Counter, Pin

def callback(counter):
    print("Counter value:", counter.value())

simple = Counter(0, pin=Pin(0, Pin.IN, pull=Pin.PULL_UP), rising=Counter.INCREMENT, threshold1=5, maximum=10, filter=1000)
simple.irq(handler=callback, trigger=Counter.IRQ_THRESHOLD1)
simple.start()

...

simple.irq(handler=callback, trigger=Counter.IRQ_THRESHOLD1, threshold1=50)

...

```
It seems that it is impossible...  :(

jonathanhogg commented:
> @jonathanhogg @harbaum Have you tried changing threshold1 like here?
> ...
> It seems that it is impossible... :(

No, you can only change `threshold1` in the `init()` method and doing so always clears the counter value back to 0. The latter appears to be a hard limitation of the hardware or driver as new threshold values only apply on a counter reset.

I think that being able to change the thresholds in the `irq()` method is confusing as it's clearly a configuration change of the unit – particularly with the implicit counter reset. However, you don't need to reset the IRQ when changing the threshold anyway, so you should be able to just swap that last line for:

```python
simple.init(threshold1=50)
```

I don't have any hardware in front of me to test with, but that's how the code was designed.

Obviously you're going to need to change that `maximum` limit if you want to have the `threshold1` actually trigger. You can safely do that in the same `init()` call if you need to change it on the fly for some reason.

IhorNehrutsa commented:
MicroPython v1.20.0-223-g4ed55102f on 2023-08-15; ESP32 module with ESP32

```
from machine import Counter, Pin

counter = Counter(0, Pin(0, Pin.IN, pull=Pin.PULL_UP))
counter.start()
```
Traceback (most recent call last):
  File "<stdin>", line 4, in <module>
AttributeError: 'Counter' object has no attribute 'start'

jonathanhogg commented:
@IhorNehrutsa Oh, sorry! I just noticed that you're using `Counter()` instead of `PCNT()`. None of what you're trying to do will work with `Counter()`.

jonathanhogg committed:
esp32/modesp32: Add new PCNT class.

Add a new `PCNT` class to the `esp32` module that provides complete,
low-level support to the ESP32 PCNT pulse counting hardware units.

jonathanhogg commented:
So I've just rebased this and tweaked the code slightly so that it will build against ESP IDF v5.0.2, which is the release that MicroPython production builds are now being made with – and therefore what 1.21 will be using.

In exciting news, ESP IDF v5.0 has a completely new PCNT driver `driver/pulse_cnt.h`! Thankfully the old API is still supported, although now deprecated. So I'm going to have to re-write this code at some point soon. I'm not planning on doing this before 1.21 comes out, though, as it'll obviously break builds against older versions.

Also, I'm not in any hurry as I'm not actually doing any projects at the moment that are using MicroPython and I have a significant amount of other work I should be doing…

vahithosan commented:
Thank you. Which esp-idf version should I use when compiling firmware?

jonathanhogg commented:
It should be compatible with all of the 4 and 5 versions. Think last time I compiled it I did so against 5.2, but I'm away from my laptop so I can't check. You may get a warning about use of the deprecated PCNT API, but it still works fine.

jonathanhogg commented:
Good news everyone! The official version of this patch just got merged as #12346 (so close to being 12345!).

I was about to close this, but I'm going to hold off until I see the whites of its eyes…

dpgeorge commented:
@jonathanhogg was there anything in here that was missed in #12346?  Any additions/changes that we should consider merging?
